File: detox/src/android/espressoapi/web/EspressoWebDetox.js
/**

	This code is generated.
	For more information see generation/README.md.
*/


function sanitize_matcher(matcher) {
  if (!matcher._call) {
    return matcher;
  }

  const originalMatcher = typeof matcher._call === 'function' ? matcher._call() : matcher._call;
  return originalMatcher.type ? originalMatcher.value : originalMatcher;
}
class EspressoWebDetox {
  static element(element, matcher) {
    function element1(matcher) {
      if (typeof matcher !== 'object' || !matcher instanceof Array) {
        throw new Error('matcher must be an array, got ' + typeof matcher);
      }

      return {
        target: element,
        method: "element",
        args: [{
          type: "Atom<List<ElementReference>>",
          value: matcher
        }]
      };
    }

    function element2(matcher, index) {
      if (typeof matcher !== 'object' || !matcher instanceof Array) {
        throw new Error('matcher must be an array, got ' + typeof matcher);
      }

      if (typeof index !== "number") throw new Error("index should be a number, but got " + (index + (" (" + (typeof index + ")"))));
      return {
        target: element,
        method: "element",
        args: [{
          type: "Atom<List<ElementReference>>",
          value: matcher
        }, {
          type: "Integer",
          value: index
        }]
      };
    }

    if (arguments.length === 1) {
      return element1.apply(null, arguments);
    }

    if (arguments.length === 2) {
      return element2.apply(null, arguments);
    }
  }

  static get(element) {
    return {
      target: element,
      method: "get",
      args: []
    };
  }

  static tap(element) {
    return {
      target: element,
      method: "tap",
      args: []
    };
  }

  static scrollToView(element) {
    return {
      target: element,
      method: "scrollToView",
      args: []
    };
  }

  static getText(element) {
    return {
      target: element,
      method: "getText",
      args: []
    };
  }

  static runScript(element, script) {
    if (typeof script !== "string") throw new Error("script should be a string, but got " + (script + (" (" + (typeof script + ")"))));
    return {
      target: element,
      method: "runScript",
      args: [script]
    };
  }

  static runScriptWithArgs(element, script, args) {
    if (typeof script !== "string") throw new Error("script should be a string, but got " + (script + (" (" + (typeof script + ")"))));

    if (typeof args !== 'object' || !args instanceof Array) {
      throw new Error('args must be an array, got ' + typeof args);
    }

    return {
      target: element,
      method: "runScriptWithArgs",
      args: [script, {
        type: "ArrayList<Object>",
        value: args
      }]
    };
  }

  static getCurrentUrl(element) {
    return {
      target: element,
      method: "getCurrentUrl",
      args: []
    };
  }

  static getTitle(element) {
    return {
      target: element,
      method: "getTitle",
      args: []
    };
  }

  static webViewInteraction(element) {
    return {
      target: element,
      method: "webViewInteraction",
      args: []
    };
  }

  static toNotExists(element) {
    return {
      target: element,
      method: "toNotExists",
      args: []
    };
  }

  static toExists(element) {
    return {
      target: element,
      method: "toExists",
      args: []
    };
  }

  static toHaveText(element, text) {
    if (typeof text !== "string") throw new Error("text should be a string, but got " + (text + (" (" + (typeof text + ")"))));
    return {
      target: element,
      method: "toHaveText",
      args: [text]
    };
  }

  static getWebView() {
    function getWebView0() {
      return {
        target: {
          type: "Class",
          value: "com.wix.detox.espresso.web.EspressoWebDetox"
        },
        method: "getWebView",
        args: []
      };
    }

    function getWebView1(matcher) {
      return {
        target: {
          type: "Class",
          value: "com.wix.detox.espresso.web.EspressoWebDetox"
        },
        method: "getWebView",
        args: [{
          type: "Invocation",
          value: sanitize_matcher(matcher)
        }]
      };
    }

    if (arguments.length === 0) {
      return getWebView0.apply(null, arguments);
    }

    if (arguments.length === 1) {
      return getWebView1.apply(null, arguments);
    }
  }

  static expect(webElement) {
    return {
      target: {
        type: "Class",
        value: "com.wix.detox.espresso.web.EspressoWebDetox"
      },
      method: "expect",
      args: [{
        type: "WebElement",
        value: webElement
      }]
    };
  }

}

module.exports = EspressoWebDetox;

File: detox/android/detox/src/main/java/com/wix/detox/espresso/web/EspressoWebDetox.java
package com.wix.detox.espresso.web;

import android.util.Log;
import android.view.View;

import androidx.test.espresso.web.model.Atom;
import androidx.test.espresso.web.model.Atoms;
import androidx.test.espresso.web.model.ElementReference;
import androidx.test.espresso.web.model.Evaluation;
import androidx.test.espresso.web.model.SimpleAtom;
import androidx.test.espresso.web.sugar.Web.WebInteraction;
import androidx.test.espresso.web.webdriver.DriverAtoms;

import org.hamcrest.Matcher;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import javax.annotation.Nullable;

import static androidx.test.espresso.web.assertion.WebViewAssertions.webMatches;
import static androidx.test.espresso.web.sugar.Web.onWebView;
import static androidx.test.espresso.web.webdriver.DriverAtoms.getText;
import static org.hamcrest.CoreMatchers.containsString;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.CoreMatchers.not;

public class EspressoWebDetox {

    private static final String TAG = "EspressoWebDetox";

    private EspressoWebDetox() {
        // static class
    }

//getWebView.element(matcher)
    public static final class WebViewElement {

        private final WebInteraction<Void> webViewInteraction;

        WebViewElement(@Nullable Matcher<View> matcher) {
            this.webViewInteraction = matcher != null ? onWebView(matcher) : onWebView();
        }

        public WebElement element(Atom<List<ElementReference>> matcher) {
            return element(matcher, 0);
        }

        public WebElement element(Atom<List<ElementReference>> matcher, int index) {
            return new WebElement(webViewInteraction, matcher, index);
        }
    }

    public static final class WebElement {
        private final WebInteraction<Void> webViewInteraction;
        private final Atom<List<ElementReference>> matcherAtom;
        private final int index;

        WebElement(WebInteraction<Void> webViewInteraction, Atom<List<ElementReference>> matcherAtom, int index) {
            this.webViewInteraction = webViewInteraction;
            this.matcherAtom = matcherAtom;
            this.index = index;
        }

        public ElementReference get() {
            List<ElementReference> elements = webViewInteraction.perform(matcherAtom).get();

            String sc = matcherAtom.getScript();
            Log.i(TAG, "matcher script: \n" + sc);

            if (elements == null || elements.size() == 0 || index >= elements.size()) {
                throw new RuntimeException("element not found ");
            }

            return elements.get(index);
        }

        public void tap() {
            webViewInteraction.withElement(get()).perform(DriverAtoms.webClick());
        }

        public boolean scrollToView() {
            return webViewInteraction.withElement(get()).perform(DriverAtoms.webScrollIntoView()).get();
        }

        public String getText() {
            return webViewInteraction.withElement(get()).perform(DriverAtoms.getText()).get();
        }

        public Evaluation runScript(String script) {
            return webViewInteraction.withElement(get()).perform(new SimpleAtom(script)).get();
        }

        public Evaluation runScriptWithArgs(String script, final ArrayList<Object> args) {
            return webViewInteraction.withElement(get()).perform(Atoms.scriptWithArgs(script, args)).get();
        }

        public String getCurrentUrl() {
            return webViewInteraction.withElement(get()).perform(Atoms.getCurrentUrl()).get();
        }

        public String getTitle() {
            return webViewInteraction.withElement(get()).perform(Atoms.getTitle()).get();
        }

    }

    public static final class ExpectElement {
        private final WebElement webElement;

        ExpectElement(WebElement webElement) {
            this.webElement = webElement;
        }

        private WebInteraction<Void> webViewInteraction() {
            return webElement.webViewInteraction;
        }

        public void toNotExists() {
            webViewInteraction().check(webMatches(webElement.matcherAtom, equalTo(Collections.<ElementReference>emptyList())));
        }

        public void toExists() {
            webViewInteraction().check(webMatches(webElement.matcherAtom, not(equalTo(Collections.<ElementReference>emptyList()))));
        }

        public void toHaveText(String text) {
            webViewInteraction().withElement(webElement.get()).check(webMatches(getText(), containsString(text)));
        }
    }

    public static WebViewElement getWebView() {
        return getWebView(null);
    }

    public static WebViewElement getWebView(@Nullable Matcher<View> matcher) {
        return new WebViewElement(matcher);
    }

    public static ExpectElement expect(WebElement webElement) {
        return new ExpectElement(webElement);
    }

}